aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -927,17 +928,19 @@ Tensor mkldnn_convolution_transpose(
   c10::MaybeOwned<Tensor> bias_maybe_owned = at::borrow_from_optional_tensor(bias_opt);
   const Tensor& bias = *bias_maybe_owned;
 
-  if (input.scalar_type() == ScalarType::BFloat16) {
+  if (input_t.scalar_type() == ScalarType::BFloat16) {
     TORCH_CHECK(mkldnn_bf16_device_check(),
         "mkldnn_convolution_transpose: bf16 path needs the cpu support avx512bw, avx512vl and avx512dq");
   }
 
-  bool use_channels_last = mkldnn_conv_use_channels_last(input, weight);
-  auto memory_format = mkldnn_convolution_memory_format(input.ndimension(), use_channels_last);
+  bool use_channels_last = mkldnn_conv_use_channels_last(input_t, weight_t);
+  auto memory_format = mkldnn_convolution_memory_format(input_t.ndimension(), use_channels_last);
 
-  auto output_sizes = conv_input_size(input.sizes(), weight.sizes(), padding, output_padding, stride, dilation, groups);
-  auto output = at::empty({0}, input.options());
+  auto output_sizes = conv_input_size(input_t.sizes(), weight_t.sizes(), padding, output_padding, stride, dilation, groups);
+  auto output = at::empty({0}, input_t.options());
 
+  auto input = input_t.is_mkldnn() ? input_t : input_t.contiguous(memory_format);
+  auto weight = weight_t.is_mkldnn() ? weight_t : weight_t.contiguous(memory_format);
```

Comment:
If input is already in channels last format while weight is not, is it better to not do channels last conversion here and fuse this with weight packing before invoking mkldnn conv primitive?

Reply:
Add contiguous for deconv here to align with the behavior of conv (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mkldnn/Conv.cpp#L223-L224). Actually, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L1583 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L2169 already do contiguous for input and weight.